pkg/controller/machinedeployment/controller_test.go

```diff
@@ -0,0 +1,896 @@
+	f.expectUpdateDeploymentStatusAction(d)
+
+	f.run(d)
+}
+
+func TestSyncDeploymentDontDoAnythingDuringDeletion(t *testing.T) {
+	f := newFixture(t)
+
+	d := newMachineDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	now := metav1.Now()
+	d.DeletionTimestamp = &now
+	f.dLister = append(f.dLister, d)
+	f.objects = append(f.objects, d)
+
+	f.run(d)
+}
+
+func TestGetMachineSetsForDeployment(t *testing.T) {
+	tests := []struct {
+		name                 string
+		noDeploymentSelector bool
+		diffCtrlRef          bool
+		diffLabels           bool
+		noCtrlRef            bool
+		expectedMachineSets  int
+	}{
+		{
+			name:                "scenario 1. machine set returned.",
+			expectedMachineSets: 1,
+		},
+		{
+			name:                "scenario 2. machine set with diff controller ref, machine set not returned.",
+			diffCtrlRef:         true,
+			expectedMachineSets: 0,
+		},
+		{
+			name:                 "scenario 3. deployment with no selector, machine set not returned.",
+			noDeploymentSelector: true,
+			expectedMachineSets:  0,
+		},
+		{
+			name:                "scenario 4. machine set with non-matching labels not returned.",
+			diffLabels:          true,
+			expectedMachineSets: 0,
+		},
+		{
+			name:                "scenario 5. machine set with no controller ref, machine set returned.",
+			noCtrlRef:           true,
+			expectedMachineSets: 1,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Log(test.name)
+			d1 := newMachineDeployment("d1", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+			d2 := newMachineDeployment("d2", 1, nil, nil, nil, map[string]string{"foo": "bar2"})
+			ms := newMachineSet(d1, "ms", 1)
+
+			if test.diffCtrlRef {
+				ms.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(d2, controllerKind)}
+			}
+			if test.noDeploymentSelector {
+				d1.Spec.Selector = metav1.LabelSelector{}
+			}
+			if test.diffLabels {
+				ms.ObjectMeta.Labels = map[string]string{"foo": "bar2"}
+			}
+			if test.noCtrlRef {
+				ms.ObjectMeta.OwnerReferences = nil
+			}
+
+			f := newFixture(t)
+
+			f.dLister = append(f.dLister, d1, d2)
+			f.msLister = append(f.msLister, ms)
+			f.objects = append(f.objects, d1, d2, ms)
+
+			c := f.newController()
+
+			msList, err := c.getMachineSetsForDeployment(d1)
+			if err != nil {
+				t.Fatalf("unexpected err calling getMachineSetsForDeployment, %v", err)
+			}
+			if test.expectedMachineSets != len(msList) {
+				t.Fatalf("got %v machine sets, expected %v machine sets, %v", len(msList), test.expectedMachineSets, msList)
+			}
+		})
+	}
+}
+
+func hasExpectedMachineNames(t *testing.T, mList *v1alpha1.MachineList, msName string, numMSMachines int) {
+	var names, expectedNames []string
+	for _, m := range mList.Items {
+		names = append(names, m.Name)
+	}
+	sort.Strings(names)
+
+	for i := 0; i < numMSMachines; i++ {
+		expectedNames = append(expectedNames, fmt.Sprintf("%v-machine-%v", msName, i))
```

Comment:
at the moment, it will probably error out or get truncated. if it is the latter, it might get confusing.
i'll file an issue to make sure we handle the case, that is check the generated name does not exceed the max length, and that the spec in deployment/machineset have base names that does is short enough to allow a hash to be appended.